wagtail/contrib/redirects/signal_handlers.py

```diff
@@ -0,0 +1,95 @@
+import logging
+
+from typing import Optional
+
+from django.conf import settings
+from django.db.models import Q
+
+from wagtail.core.models import BaseLogEntry, Page, PageBase
+from wagtail.core.utils import BatchCreator, get_dummy_request
+
+from .models import Redirect
+
+
+logger = logging.getLogger(__name__)
+
+
+class BatchRedirectCreator(BatchCreator):
+    """
+    A specialized ``BatchCreator`` class for saving ``Redirect`` objects.
+    """
+
+    model = Redirect
+
+    def initialize_instance(self, kwargs):
+        # Automatically normalise paths when creating instances from kwargs
+        kwargs["old_path"] = Redirect.normalise_path(kwargs["old_path"])
+        return super().initialize_instance(kwargs)
+
+    def pre_process(self):
+        # delete any existing automatically-created redirects that might clash
+        # with the items in `self.items`
+        clashes_q = Q()
+        for item in self.items:
+            clashes_q |= Q(old_path=item.old_path, site_id=item.site_id)
+        Redirect.objects.filter(automatically_created=True).filter(clashes_q).delete()
+
+
+def autocreate_redirects(
+    sender: PageBase, instance: Page, url_path_before: str, url_path_after: str, log_entry: Optional[BaseLogEntry] = None
+) -> None:
+    if instance.is_site_root():
+        # url_path changes for site root pages do not affect URLs generally,
+        # so we can exit early here.
+        return None
+
+    # This shouldn't ever be the case, but...
+    assert url_path_before != url_path_after
```

Comment:
We should avoid using assertions as they throw exceptions, and creating redirects doesn't feel like a critical part of the request that's worth crashing over.

Could we return and log a warning instead?

Also, the ``assert`` statement can be removed if Python is being run with optimisations enabled.

Reply:
I suppose we can remove this completely, really. That'll keep things nice and tidy.